### ISSUE TITLE
uri?add-to-cart=<productId> url is adding product twice to cart if url is redirected

### ISSUE THREAD

**Issue by alapiere** [woothemes/woocommerce#52630]
### Prerequisites

- [x] I have carried out troubleshooting steps and I believe I have found a bug.
- [x] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

When querying /checkout?add-to-cart=123 , product is added twice to cart instead of once

### Expected behavior

Product must be added only once even when url is redirected,
it must not be added to cart on redirected url

### Actual behavior

When querying /checkout?add-to-cart=123 ,
I'm redirected 301 to the correct url which is /checkout/?add-to-cart=123 ( "/" after checkout)

Product is added to cart on first 301 hit,
then again on the correct url,
--> leaving the product twice in cart

### Steps to reproduce

- go to /checkout?add-to-cart=123 (replace 'checkout' by your checkout slug if it's different)
- user is redirected to /checkout/?add-to-cart=123 
- product is added twice to cart

### WordPress Environment

<details><summary>System Status Report</summary>

```
### WordPress Environment ###

WordPress address (URL): [Redacted]
Site address (URL): [Redacted]
WC Version: 9.3.3
Legacy REST API Package Version: ✔ 1.0.4
Action Scheduler Version: ✔ 3.8.1
Log Directory Writable: ✔
WP Version: 6.6.2
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: –
WP Cron: ✔
Language: fr_FR
External object cache: –

### Server Environment ###

Server Info: nginx/1.25.5
PHP Version: 8.1.30
PHP Post Max Size: 128 MB
PHP Time Limit: 300
PHP Max Input Vars: 10000
cURL Version: 7.68.0
OpenSSL/1.1.1f

SUHOSIN Installed: –
MySQL Version: 10.11.5-MariaDB-1:10.11.5+maria~ubu2004
Max Upload Size: 128 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

[REDACTED]

### Post Type Counts ###

acf-field: 83
acf-field-group: 13
acf-post-type: 4
acf-taxonomy: 2
apto_sort: 8
artist: 159
attachment: 4402
custom_css: 1
customize_changeset: 76
exhibition_on_site: 82
exhibition_to_rent: 109
faq: 102
nav_menu_item: 19
page: 40
polylang_mo: 2
post: 22
product: 5449
product_variation: 170
revision: 1638
shop_coupon: 14
shop_order: 1358
shop_order_refund: 9
slider: 38
wcp_ruleset: 1
wp_block: 1
wp_global_styles: 1

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔

### Active Plugins (3) ###

Advanced Custom Fields PRO: by WP Engine – 6.2.1.1
WooCommerce Legacy REST API: by WooCommerce – 1.0.4
WooCommerce: by Automattic – 9.3.3

### Inactive Plugins (0) ###


### Must Use Plugins (2) ###

disable-admin-notif-user-change-pwd.php: by  –
Kinsta Must-use Plugins: by Kinsta Team – 2.4.8

### Settings ###

Legacy API Enabled: ✔
Force SSL: –
Currency: EUR (€)
Currency Position: right
Thousand Separator: 
Decimal Separator: ,
Number of Decimals: 2
Taxonomies: Product Types: external (external)
gift-card (gift-card)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: –
Enforce Approved Product Download Directories: –
HPOS feature enabled: –
Order datastore: WC_Order_Data_Store_CPT
HPOS data sync enabled: –

### Logging ###

Enabled: ✔
Handler: Automattic\WooCommerce\Internal\Admin\Logging\LogHandlerFileV2
Retention period: 30 days
Level threshold: –
Log directory size: 579 KB

### WC Pages ###

Shop base: #67 - /oeuvres-en-vente/
Cart: #68 - /cart/ -  Contains the [woocommerce_cart] shortcode
Checkout: #69 - /checkout/ -  Contains the [woocommerce_checkout] shortcode
My account: #70 - /mon-compte/
Terms and conditions: #18 - /conditions-generales-vente/

### Theme ###

Name: Gallerie Robillard
Version: 99.0
Author URL: 
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ✔

### Templates ###

Archive Template: Your theme has a woocommerce.php file
you will not be able to override the woocommerce/archive-product.php custom template since woocommerce.php has priority over archive-product.php. This is intended to prevent display issues.

Overrides: robillard/woocommerce/single-product/add-to-cart/simple.php

### Admin ###

Enabled Features: activity-panels
analytics
product-block-editor
coupons
core-profiler
customize-store
customer-effort-score-tracks
import-products-task
experimental-fashion-sample-products
shipping-smart-defaults
shipping-setting-tour
homescreen
marketing
mobile-app-banner
navigation
onboarding
onboarding-tasks
pattern-toolkit-full-composability
product-custom-fields
remote-inbox-notifications
remote-free-extensions
payment-gateway-suggestions
printful
shipping-label-banner
subscriptions
store-alerts
transient-notices
woo-mobile-welcome
wc-pay-promotion
wc-pay-welcome-page
launch-your-store

Disabled Features: experimental-blocks
minified-js
product-pre-publish-modal
settings
async-product-editor-category-field
product-editor-template-system
blueprint
reactify-classic-payments-settings

Daily Cron: ✔ Next scheduled: 2024-11-07 08:26:54 +02:00
Options: ✔
Notes: 81
Onboarding: completed

### Action Scheduler ###

Complete: 57
Oldest: 2024-10-30 18:51:38 +0000
Newest: 2024-11-06 17:49:22 +0000

Failed: 1,801
Oldest: 2023-10-27 00:05:15 +0000
Newest: 2024-11-06 17:49:18 +0000

Pending: 1
Oldest: 2024-11-07 17:41:17 +0000
Newest: 2024-11-07 17:41:17 +0000


### Status report information ###

Generated at: 2024-11-06 19:49:44 +02:00
```
</details>

### Isolating the problem

- [x] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/products/storefront/).
- [x] I can reproduce this bug consistently using the steps above.

**Comment by alapiere** [comment]
Issue is reproduced by wp.org woocommerce support team : 
https://wordpress.org/support/topic/woocommerce-adds-2-products-to-cart-because-of-redirect/
With the precision : 
"it looks like this behavior appears only when Default customer location is set to Geolocate with page caching support and Add to cart behaviors are unchecked."

I have indeed the "redirect" disabled, not the ajax action. (and no clue about this "geolocate" option)

Note that it was also reported in SO 6 years ago : https://stackoverflow.com/questions/49699928/custom-add-to-cart-url-adding-product-twice-to-the-cart-in-woocommerce-3-3-1

**Comment by senadir** [comment]
I can reproduce on my stock setup, no geolocation needed.


Whatever is doing the add to cart needs to be a bit late to account for redirect, otherwise the query should be deleted.

**Comment by senadir** [comment]
The issue here is that `WC_Form_Handler::add_to_cart_action` is running too early in `wp_loaded`, while all redirects happen on `template_redirect`.

A solution would be to switch `add_to_cart_action` and `update_cart_action` to `template_redirect`.

Doing that alone, and keeping priority on 20, would surface another issue in which doing something like `checkout?add-to-cart=27` on an empty cart would result in the item not being added and a redirect to cart.

This is because the logic that redirects empty cart is on `wc_template_redirect` is at `template_redirect:10`, so that one should be pushed further, but no further than `add_to_cart_action`, so the sequence should be:

- `canonical_redirect` from WP, happens first in `template_redirect` this should fix the `/` issue. at 10.
- `add_to_cart_action` from WC happens at `template_redirect` at 15.
- `wc_template_redirect` from WC happens at `template_redirect` at 20.

The risk is that if something is relying on running after those things might break.

Another solution would be to keep `add_to_cart_action` at 10 and keep `wc_template_redirect` but this risks race condition, though to me it's unlikely given how the code is structured: WP code will always register first, `class-wc-form-handler.php` is loaded after that, in `load-post.php` hook, and `wc_template_redirect` is in `after_setup_theme`.

**Comment by pmcpinto** [comment]
Thanks for the report. I know this has been open a while, and I appreciate your patience. We're currently going through some older WooCommerce issues, identifying what still needs fixing, and tidying up what's already fixed or out of scope.

I've taken a look and marked this ready for the engineering team to pick up.
